features/netsocket/nsapi_types.h

```diff
@@ -48,6 +48,9 @@ enum nsapi_error {
     NSAPI_ERROR_DHCP_FAILURE  = -3010,     /*!< DHCP failed to complete successfully */
     NSAPI_ERROR_AUTH_FAILURE  = -3011,     /*!< connection to access point failed */
     NSAPI_ERROR_DEVICE_ERROR  = -3012,     /*!< failure interfacing with the network processor */
+    NSAPI_ERROR_IN_PROGRESS   = -3013,     /*!< operation (eg connect) in progress */
```

Comment:
As I understand it, the distinctions seem to be:

"would block" - "that call had no effect, but you can try the same thing again later and in might work"
"in progress"- "that call started something, but it hasn't finished, check back for status later"
"already" - "that call was trying to start something already started, so it was ignored"

In the case of "connect", checking back after "in progress" could either be trying connect again (in which case you can expect "already in progress" or "is connected" errors), or just going ahead and trying to read/write (in which case you'd get "not connected" if the connect hadn't completed yet).